### DISCUSSION TITLE
Moving to docker

### DISCUSSION THREAD

**Discussion by Vibhigupta** [mastodon/mastodon#27394] [General]
Hello, I am thinking to move my instance to docker but i dont know, how to migrate database and make it use s3? i was thinking to make changes to docker file. 

can someone help me with this? what changes does i have to make in the dockerfile and dockercompose?

**Comment by winteriscariot** [comment]
i as well would like some info. in particular, the .env.production.sample file included from this repo is not configured properly for dockerized deployment and will fail to connect to the services -- redis, postgres -- because .env.production assumes you're running it on a host and not in containers. I unfortunately do not know enough about Docker to be able to properly construct a new .env.production file, and it appears that the official Mastodon docker documentation (https://github.com/tootsuite/documentation/blob/master/Running-Mastodon/Docker-Guide.md) has vanished (it doesn't appear on docs.joinmastodon.com)

**Comment by Vibhigupta** [comment]
@winteriscariot  Thank You for the comment

I was thinking to replace the env file with mine own but database migration is something I'm worried about. i will share my finding after testing on local machine

**Comment by winteriscariot** [comment]
So I just spent the past hour or so working on this and I was successful in getting it online (thanks to a friend of mine who knows docker well, and he's taught me a lot about Docker and Compose over the past hour). There are 3 issues:

1. The .env.production.sample file (which you should cp to .env.production) is not configured for containerized operation, but this is easily changed. The first issue is that `REDIS_HOST` and `DB_HOST` are set to host variables: these need to be changed to docker hostnames that they can connect to. You'll want the options changed as follows, in your .env.production file:
  
```
    REDIS_HOST=redis

    DB_HOST=db
```

The default options will cause connections to fail, since the containers can't connect to the host sockets or addresses. (127.0.0.1 doesn't mean anything to a container.) Since the redis and db containers are using internal docker networking, they can be resolved using their service names (ie, `db` and `redis`)
 
2. The other issue is that the db doesn't get properly initialized, and so the `docker-compose run --rm web rails db:migrate` command to populate the db will fail. This will cause the db container to loop, and you'll get errors that none of the other containers can connect to it. In order to fix this, create a new file in your working `mastodon` directory called .env.db, and add the following lines to it:

```
    POSTGRES_PASSWORD=<secret password>
    POSTGRES_DB=mastodon_production
    POSTGRES_USER=mastodon
```

Note: these should match the `DB_PASS`, `DB_NAME`, and `DB_USER` from your .env.production file. Then add the following line to your `db` service in docker-compose.yml (i put it between the `image` and `shm_size` options though it doesn't particularly matter):

`    env_file: .env.db`

What this does is ensure that the `postgres` container will properly create the db and assign the user and password to it. Once you've done this, you can run `docker-compose build` for one last time.

After the build completes, run the following command: 

`    docker-compose run --rm web rails db:migrate`

this will populate the database and make sure you're not connecting to an empty database. Once this completes, you should be able to successfully run `docker-compose up -d` and it will start. (Note: this assumes you've edited your .env.production in all other ways, including changing your domain name and updating secret keys).

I'll be around if you have any questions!

**Comment by Vibhigupta** [comment]
@winteriscariot  Thank you ,

can you help me in one more thing, I want to basically migrate my already running instance so how do i migrate the db, eg you created another file to make db initialize properly but i already have db ( backup.dump) so can you ask your friend how do i import my database. 

I am guessing i have to to copy the dump file into the dir and run the import cmd through docker compose. will that work?

https://docs.joinmastodon.org/admin/migrating/
